### PR TITLE
Subforms: Implementing per-field filters and validation.

### PR DESCRIPTION
**Pull request by Harmageddon**
This pull request introduces the option to both filter and validate subforms field by field. The intention is that every single field inside a subform can have its own filter and validation parameters. Before this PR, these parameters would just be ignored.

### Summary of Changes
* Introduced a Parameter `filter="subform"` to filter a subform field by filtering its child fields individually.
* Introduced a Parameter `validate="subform"` to validate a subform field by validating its child fields individually.
* Moved the loading of the JForm instance and its data inside the subform field from the `getInput` method into two separate methods to avoid repeated code, because the subform has to be loaded for filtering and validation, too. This change should **not** change any present behavior of subform fields, because I only moved the existing code into two additional methods.

### Testing Instructions
* Test with a form containing a subform field.
  * I created a small module for testing purposes on [mod_demovalidation](https://github.com/Harmageddon/mod_demovalidation). It has no frontend, only the form in the module configuration in backend. Feel free to use this, but I'd recommend that you play around with different parameter combinations to not just replicate my tests.
  * You can also use the testing instructions from #7829 and add some filters and validations...
  * ... or those from #12813, where an inline subform is used.
  * If you are developing an extension using subforms, this is the perfect opportunity to improve it! ;-)
* Test the impact of filters.
  * Add `filter="subform"` to the subform field.
  * For example, add `filter="int"` to a field inside the subform XML and try to save values like "abc" (should be filtered to 0) or "0002" (should be filtered to 2).
  * Make sure that correct values are not filtered and everything is stored correctly.
* Test the impact of validation.
  * Add `validate="subform"` to the subform field.
  * For example, add `validate="email"` to a field inside the subform XML and try to save values like "foo".
  * Make sure that correct values are not marked as invalid and everything is stored correctly.
* Make sure that the subform field behaves like before (apart from the added filtering and validation options).
* Test everything with `multiple="true"`, `multiple="false"` and without the `multiple` parameter for the subform field.

### Expected result
* Unfiltered values should be filtered, depending on the filters of the fields inside the subform XML.
* Invalid values should be rejected, depending on the validation rules of the fields inside the subform XML.

If a validation fails, there are three options what to display as an error message:
1. "Invalid Field: Subform" (Title of the subform field containing at least one invalid value)
2. "Invalid Field: Mail" (Title of the first invalid single field inside the subform)
3. Like 2., but for all invalid fields.

I used option 2, with a fallback to 1. For a subform containing invalid values, only one warning message will be returned, containing the title of the first field with an invalid value. I'm preferring this over option 1, because it is more specific about where the error is. Option 3 would be possible using additional messages, but as the basic implementation of validation itself displays a maximum of 3 messages, I thought it would be sufficient to send only one here.

### Documentation Changes Required
* ~As I have no experience in the version numbering of the CMS libraries, I'd be thankful if someone could tell me which value should be added for the `@since` tag for the `JFormRuleSubform` class.~
* When the pull request is merged and released, I will add short documentation about the additional parameters on docs.joomla.org.

### CODE REVIEW

**Comment by zero-24** [comment]
@Harmageddon with the last 3 commit i have just fixed a error from drone and minimal cs ;)

**Comment by Harmageddon** [comment]
@zero-24 Awesome, thank you! Didn't know about this makro, very useful!

**Comment by Harmageddon** [comment]
Added some small code cleanups, inspired by @matrikular. Thank you!

**Comment by laoneo** [comment]
Nice pr. Can you please fix the merge conflicts that we can test it?

**Comment by Harmageddon** [comment]
@laoneo Thank you! It should work now.

File: libraries/joomla/form/fields/subform.php [hunk 1]
```diff
@@ -356,4 +327,74 @@ protected function getName($fieldName)
 
 		return $name;
 	}
+
+	/**
+	 * Loads the form instance for the subform.
+	 *
+	 * @return  JForm  The form instance.
```
Comment (laoneo) [inline libraries/joomla/form/fields/subform.php]: Can you use the namespaced class? Thanks.

File: libraries/joomla/form/fields/subform.php [hunk 2]
```diff
@@ -356,4 +327,74 @@ protected function getName($fieldName)
...
+		$control = $this->name;
+
+		if ($this->multiple)
+		{
+			$control .= '[' . $this->fieldname . 'X]';
+		}
+
+		// Prepare the form template
+		$formname = 'subform.' . ($this->group ? $this->group . '.' : '') . $this->fieldname;
+		$tmpl     = JForm::getInstance($formname, $this->formsource, array('control' => $control));
```
Comment (laoneo) [inline libraries/joomla/form/fields/subform.php]: Can you use the namespaced class? Thanks.

File: libraries/joomla/form/fields/subform.php [hunk 3]
```diff
@@ -356,4 +327,74 @@ protected function getName($fieldName)
...
+
+		return $tmpl;
+	}
+
+	/**
+	 * Binds given data to the subform and its elements.
+	 *
+	 * @param   JForm  &$subForm  Form instance of the subform.
```
Comment (laoneo) [inline libraries/joomla/form/fields/subform.php]: Can you use the namespaced class? Thanks.

File: libraries/joomla/form/fields/subform.php [hunk 4]
```diff
@@ -356,4 +327,74 @@ protected function getName($fieldName)
...
+	 *
+	 * @return  array  Array of JForm instances for the rows.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	private function loadSubFormData(&$subForm)
```
Comment (laoneo) [inline libraries/joomla/form/fields/subform.php]: Can you type hint it?

File: libraries/joomla/form/fields/subform.php [hunk 5]
```diff
@@ -356,4 +327,74 @@ protected function getName($fieldName)
...
+		$forms = array();
+		$value = array_values($value);
+
+		// Show as many rows as we have values, but at least min and at most max.
+		$c = max($this->min, min(count($value), $this->max));
+
+		for ($i = 0; $i < $c; $i++)
+		{
+			$control  = $this->name . '[' . $this->fieldname . $i . ']';
+			$itemForm = JForm::getInstance($subForm->getName() . $i, $this->formsource, array('control' => $control));
```
Comment (laoneo) [inline libraries/joomla/form/fields/subform.php]: Can you use the namespaced class? Thanks.

File: libraries/joomla/form/rule/subform.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+defined('JPATH_PLATFORM') or die;
+
+use Joomla\Registry\Registry;
+
+/**
+ * Form rule to validate subforms field-wise.
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class JFormRuleSubform extends JFormRule
```
Comment (laoneo) [inline libraries/joomla/form/rule/subform.php]: Can you use the namespaced class? Thanks.

File: libraries/joomla/form/rule/subform.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	 *                                      For example if the field has name="foo" and the group value is set to "bar" then the
+	 *                                      full field name would end up being "bar[foo]".
+	 * @param   Registry          $input    An optional Registry object with the entire data set to validate against the entire form.
+	 * @param   JForm             $form     The form object for which the field is being tested.
+	 *
+	 * @return  boolean  True if the value is valid, false otherwise.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function test(SimpleXMLElement $element, $value, $group = null, Registry $input = null, JForm $form = null)
```
Comment (laoneo) [inline libraries/joomla/form/rule/subform.php]: Can you use the namespaced class? Thanks.

File: libraries/joomla/form/rule/subform.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	/**
+	 * Method to test given values for a subform..
+	 *
+	 * @param   SimpleXMLElement  $element  The SimpleXMLElement object representing the `<field>` tag for the form field object.
+	 * @param   mixed             $value    The form field value to validate.
+	 * @param   string            $group    The field name group control value. This acts as as an array container for the field.
+	 *                                      For example if the field has name="foo" and the group value is set to "bar" then the
+	 *                                      full field name would end up being "bar[foo]".
+	 * @param   Registry          $input    An optional Registry object with the entire data set to validate against the entire form.
+	 * @param   JForm             $form     The form object for which the field is being tested.
```
Comment (laoneo) [inline libraries/joomla/form/rule/subform.php]: Can you use the namespaced class? Thanks.

**Review by laoneo — CHANGES_REQUESTED** [review]

**Comment by laoneo** [comment]
Thanks, added a few comments to use the namespaced classes. If done I will test it.

**Comment by Harmageddon** [comment]
Done. Thanks in advance!

**Comment by laoneo** [comment]
Now you got me wrong, use the real classes like \Joomla\CMS\Form\Form, etc. Not just adding a backslash before them.

**Comment by Harmageddon** [comment]
Oh, sorry. Should I copy the rule to libraries/src/Form/Rule, too, then? Or does it work with this "mixed" form (the subform field and the rule in libraries/joomla/form and the form and the JFormFieldRule class in libraries/src/Form)?

**Comment by laoneo** [comment]
Yes please copy the rule also to `libraries/src/Form/Rule`. The field has to stay because of autoloading. In Joomla 4 we have moved it also to `libraries/src/Form/Field`.

**Comment by Harmageddon** [comment]
Okay, a little bit confusing. I hope I haven't messed up anything.

File: libraries/joomla/form/rule/subform.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
+<?php
```
Comment (laoneo) [inline libraries/joomla/form/rule/subform.php]: This one can be deleted.

Comment (Harmageddon) [reply libraries/joomla/form/rule/subform.php]: The `<?php` tag? Why?

Comment (laoneo) [reply libraries/joomla/form/rule/subform.php]: I meant the whole file as you have the namespaced one.

Comment (Harmageddon) [reply libraries/joomla/form/rule/subform.php]: Ah, I see. Seems I have missed that the rules have been moved completely.

File: libraries/joomla/form/fields/subform.php [hunk 10]
```diff
@@ -7,6 +7,8 @@
  * @license     GNU General Public License version 2 or later; see LICENSE
  */
 
+use Joomla\CMS\Form\Form;
```
Comment (laoneo) [inline libraries/joomla/form/fields/subform.php]: Minor cosmetics here, it belongs below the defined check.

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on d379667af4c8f474d11788512e766dc0de723be3<br /><br />It worked. I'v tested it by changing the file plugins/fields/checkboxes/params/checkboxes.xml to
```
<?xml version="1.0" encoding="utf-8"?>
<form>
	<fields name="fieldparams">
		<fieldset name="fieldparams">
			<field
				name="options"
				type="subform"
				label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_LABEL"
				description="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_DESC"
				layout="joomla.form.field.subform.repeatable-table"
				icon="list"
				multiple="true"
				filter="subform"
				validate="subform"
			>
				<form hidden="true" name="list_templates_modal" repeat="true">
					<field
						name="name"
						type="text"
						label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_NAME_LABEL"
						size="30"
					/>

					<field
						name="value"
						type="text"
						label="PLG_FIELDS_CHECKBOXES_PARAMS_OPTIONS_VALUE_LABEL"
						size="30"
						validate="color"
					/>
				</form>
			</field>
		</fieldset>
	</fields>
</form>
```

After that I created a checkbox field and tried to create an option where the value is not a hex number. It failed. The I changed `validate="color"`to `filter="int"` and tried to save again, then it successfully saved it but turned the value into an integer.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14189">issues.joomla.org/tracker/joomla-cms/14189</a>.</sub>

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on 6b79ceab71790c031b975abea4efbe1ebc15dbc5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14189">issues.joomla.org/tracker/joomla-cms/14189</a>.</sub>

**Comment by ghost** [comment]
@laoneo can you please retest?

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 6b79ceab71790c031b975abea4efbe1ebc15dbc5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14189">issues.joomla.org/tracker/joomla-cms/14189</a>.</sub>

**Review by laoneo — APPROVED** [review]

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by laoneo** [comment]
Perhaps you should consider to make your own callback for filter instead of hardcoding the subform filter into the form class. More details can be found here https://github.com/joomla/joomla-cms/issues/20569#issuecomment-391689844.

**Comment by Harmageddon** [comment]
Are there any other filters that could serve as an example? I assume that I'd have to create a new class? Where should this class be located? As far as I see, most core rules (like "access") are directly in the Form class. 

I would use the method you suggest only for third-party extensions and keep the core filters in one place - either in the Form class as of now or each one in a dedicated class. If we keep some filters in one and others in another place, we risk inconsistencies (i.e. if the filter mechanism gets changed) and people will have a harder time to find all available filters when digging through the code. Personally, I'd support your approach to create classes for filters, because I remember debugging the filter method until I found the hardcoded filters in the Form class. But I'd prefer to either move all or none.

**Comment by mbabker** [comment]
What is hardcoded into the Form class' `filterField` method is really a bunch of special cases specific to the form data that don't belong to the InputFilter class.  So, "core filters" don't need to be hardcoded into the Form class itself unless there is a strong reason why it must be part of the Form API and available at all times.

**Comment by Harmageddon** [comment]
Okay... So is there any best practice where else to put such a filter? And if I see this right, we'd then require users/developers to use a `filter="SomeHelperClass:subform"` instead of just `filter="subform"`?

**Comment by SharkyKZ** [comment]
IMO, form field filtering and validation should be implicit whenever possible to reduce XML form maintenance needs.

So `Form`'s `validateField` and `filterField` logic could be moved to `FormField`. Subform field could then override parent methods and other fields could declare default filters and validation rules to be used when none is added in XML.

Here's a mockup based on your code and #12414 https://github.com/SharkyKZ/joomla-cms/commit/fa514ae38f3a534a0afea2c7f454554f69c4a414.

**Comment by mbabker** [comment]
> So Form's validateField and filterField logic could be moved to FormField. Subform field could then override parent methods and other fields could declare default filters and validation rules to be used when none is added in XML.

Not necessarily, because each part of the API has a separate concern.

`FormField` is in essence the visual representation of the field and handles the rendering, the class chain does not belong in the filtering/validation process at all.

`FormRule` defines some validation rules to be applied when validating the form's data, these are called during the validation process only.

`Form` is the data manager for a form instance and this is where the filtering and validation processes belong.

**Comment by SharkyKZ** [comment]
That's the way it is now but is it really correct? 

>FormField is in essence the visual representation of the field and handles the rendering, the class chain does not belong in the filtering/validation process at all.

Form fields also provide the data options and data type. So filtering and validation makes sense here. Filtering and validation is applied per-field anyways. That said, shouldn't `FormRule` be `FormFieldRule` in the first place?

List-based fields should be validated against options (except when custom option is allowed), numeric fields should be filtered accordingly and Subform should not accept other filters but its own (currently, adding a filter like `string` to subform field causes fatal error).

I think this functionality is vital and should be implemented one way or another.

**Comment by mbabker** [comment]
Rewrite the entire form package from scratch then?  There is already a separation of concern regarding each of the object types in the package and an established pattern of use, if you feel this is inefficient feel free to propose a new package but the existing class chain can't just be changed out of the blue like this.  By design, filtering and validation are not the responsibility of the `FormField` classes and refactoring to move responsibilities like this is no small thing.  At the end of the day though, regardless of the class structure, what is responsible for rendering should be a different aspect of the API than what is responsible for data validation (if you look at any proper framework you'll find they have a standalone validation component that the form data is pushed through whereas the closest thing we have to that is the `FormRule` class and that is tightly coupled to the `Form` class) and if we're going to start talking refactoring the form API then I don't think we should be making one sort of "super" object that does all the things.

**Comment by roland-d** [comment]
@mbabker Is this something for 3.9?

**Comment by Harmageddon** [comment]
So with this PR being marked as RTC for almost 4 months now, I'd like to ask what is the status for this one? Is it going to be merged for some future release in the current state or do you prefer me to move the filter to some different location for architectural reasons?
@mbabker @wilsonge (pinging you as the release leads for 3.9/3.10)

**Comment by mbabker** [comment]
I'm hesitant to merge this because I don't think it is architecturally done correctly and inherently once it ships changing the API will never happen.

**Comment by Harmageddon** [comment]
I see. However, the problem with moving this method outside the `Form` class is, that we need the form field (`$element`) in order to load the subform and get the filters for the respective fields inside the subform. Callback filters as well as the `InputFilter->clean` function only get the value of the field as an argument, but not the whole field.
So a callback filter would get an array containing the values of the fields inside the subform without knowing which value to treat with which filter.

**Comment by Harmageddon** [comment]
Merged staging to include the changes from #17552.

@mbabker With the current API, I don't see any alternative to this implementation.

**Comment by Harmageddon** [comment]
(sorry, I accidentally hit the wrong button)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14189">issues.joomla.org/tracker/joomla-cms/14189</a>.</sub>

**Comment by Hackwar** [comment]
Please see #12414. That might have an impact on this PR.

**Comment by Harmageddon** [comment]
Thank you for your work there! It does indeed have an impact. I'm going to rewrite this PR for 4.0 then. And as this PR here probably won't be ever merged into 3.x after being RTC for 8 months, I'm now closing it and hope for better luck and API integration for 4.0.

**Comment by wilsonge** [comment]
Just for the record I'd definitely be interested in seeing this for Joomla 4!

**Comment by Harmageddon** [comment]
Apparently, the changes proposed in this PR have been added in b06019e83efc67e06e694753ab114d7c98eaf5cf and d34a3ec, so I assume they're going to be merged into 4.0, too, and there is no need for a further PR? 
@HLeithner / @wilsonge what do you think?

**Comment by zero-24** [comment]
Correct. All 3.x changes are merged up to 4.0 before the final release.

**Pull request closed without merging**